src/wp-includes/block-patterns.php

```diff
@@ -376,116 +376,144 @@ function _register_theme_block_patterns() {
 	$themes[] = wp_get_theme( $template );
 
 	foreach ( $themes as $theme ) {
-		$dirpath = $theme->get_stylesheet_directory() . '/patterns/';
-		if ( ! is_dir( $dirpath ) || ! is_readable( $dirpath ) ) {
-			continue;
+		$transient_name = 'wp_theme_patterns_' . $theme->get_stylesheet();
+		$dirpath        = $theme->get_stylesheet_directory() . '/patterns/';
+
+		$can_use_cached = ! wp_is_development_mode( 'theme' );
+		$version        = $theme->get( 'Version' );
+
+		$pattern_datas = false;
+		if ( $can_use_cached ) {
+			$pattern_datas = get_transient( $transient_name );
+			if ( is_array( $pattern_datas ) && $pattern_datas['version'] !== $version ) {
+				$pattern_datas = false;
+			}
```

Comment:
What if it's something other than `array` or `false`? Should we be safe here and do something like this?

```php
if ( empty( $pattern_datas['version'] ) || $pattern_datas['version'] !== $version ) {
```